File: services/java-microservices/cart-service/src/main/java/com/ecommerce/cartservice/controller/CartController.java
package com.ecommerce.cartservice.controller;

import com.ecommerce.cartservice.dto.*;
import com.ecommerce.cartservice.service.CartService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.UUID;

@RestController
@RequestMapping("/api/v1/carts")
@RequiredArgsConstructor
@Slf4j
@Tag(name = "Cart", description = "Shopping Cart Management API")
public class CartController {

    private final CartService cartService;

    @PostMapping
    @Operation(summary = "Create a new cart", description = "Creates a new shopping cart for a user")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "201", description = "Cart created successfully"),
            @ApiResponse(responseCode = "400", description = "Invalid request")
    })
    public ResponseEntity<CartResponse> createCart(@Valid @RequestBody CreateCartRequest request) {
        log.info("POST /api/v1/carts - Creating cart for userId: {}", request.getUserId());
        CartResponse response = cartService.createCart(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping("/{id}")
    @Operation(summary = "Get cart by ID", description = "Retrieves a cart by its unique identifier")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Cart found"),
            @ApiResponse(responseCode = "404", description = "Cart not found"),
            @ApiResponse(responseCode = "410", description = "Cart expired")
    })
    public ResponseEntity<CartResponse> getCart(
            @Parameter(description = "Cart ID") @PathVariable UUID id) {
        log.info("GET /api/v1/carts/{}", id);
        CartResponse response = cartService.getCart(id);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/user/{userId}")
    @Operation(summary = "Get active cart for user", description = "Retrieves the active cart for a specific user")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Active cart found"),
            @ApiResponse(responseCode = "404", description = "No active cart found for user")
    })
    public ResponseEntity<CartResponse> getActiveCartForUser(
            @Parameter(description = "User ID") @PathVariable UUID userId) {
        log.info("GET /api/v1/carts/user/{}", userId);
        CartResponse response = cartService.getActiveCartForUser(userId);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/{id}/items")
    @Operation(summary = "Add item to cart", description = "Adds a product item to the specified cart")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Item added successfully"),
            @ApiResponse(responseCode = "400", description = "Invalid request"),
            @ApiResponse(responseCode = "404", description = "Cart not found"),
            @ApiResponse(responseCode = "410", description = "Cart expired")
    })
    public ResponseEntity<CartResponse> addItem(
            @Parameter(description = "Cart ID") @PathVariable UUID id,
            @Valid @RequestBody AddItemRequest request) {
        log.info("POST /api/v1/carts/{}/items - Adding productId: {}", id, request.getProductId());
        CartResponse response = cartService.addItem(id, request);
        return ResponseEntity.ok(response);
    }

    @PutMapping("/{id}/items/{itemId}")
    @Operation(summary = "Update item quantity", description = "Updates the quantity of an item in the cart")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Item updated successfully"),
            @ApiResponse(responseCode = "400", description = "Invalid request"),
            @ApiResponse(responseCode = "404", description = "Cart or item not found")
    })
    public ResponseEntity<CartResponse> updateItemQuantity(
            @Parameter(description = "Cart ID") @PathVariable UUID id,
            @Parameter(description = "Cart Item ID") @PathVariable UUID itemId,
            @Valid @RequestBody UpdateItemRequest request) {
        log.info("PUT /api/v1/carts/{}/items/{} - Updating quantity to: {}", id, itemId, request.getQuantity());
        CartResponse response = cartService.updateItemQuantity(id, itemId, request);
        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/{id}/items/{itemId}")
    @Operation(summary = "Remove item from cart", description = "Removes an item from the cart")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Item removed successfully"),
            @ApiResponse(responseCode = "404", description = "Cart or item not found")
    })
    public ResponseEntity<CartResponse> removeItem(
            @Parameter(description = "Cart ID") @PathVariable UUID id,
            @Parameter(description = "Cart Item ID") @PathVariable UUID itemId) {
        log.info("DELETE /api/v1/carts/{}/items/{}", id, itemId);
        CartResponse response = cartService.removeItem(id, itemId);
        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "Clear cart", description = "Removes all items from the cart")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "204", description = "Cart cleared successfully"),
            @ApiResponse(responseCode = "404", description = "Cart not found")
    })
    public ResponseEntity<Void> clearCart(
            @Parameter(description = "Cart ID") @PathVariable UUID id) {
        log.info("DELETE /api/v1/carts/{}", id);
        cartService.clearCart(id);
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/{id}/coupon")
    @Operation(summary = "Apply coupon to cart", description = "Applies a coupon code to the cart for a discount")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Coupon applied successfully"),
            @ApiResponse(responseCode = "400", description = "Invalid coupon code"),
            @ApiResponse(responseCode = "404", description = "Cart not found")
    })
    public ResponseEntity<CartResponse> applyCoupon(
            @Parameter(description = "Cart ID") @PathVariable UUID id,
            @Valid @RequestBody ApplyCouponRequest request) {
        log.info("POST /api/v1/carts/{}/coupon - Applying code: {}", id, request.getCouponCode());
        CartResponse response = cartService.applyCoupon(id, request);
        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/{id}/coupon")
    @Operation(summary = "Remove coupon from cart", description = "Removes the applied coupon from the cart")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Coupon removed successfully"),
            @ApiResponse(responseCode = "404", description = "Cart not found")
    })
    public ResponseEntity<CartResponse> removeCoupon(
            @Parameter(description = "Cart ID") @PathVariable UUID id) {
        log.info("DELETE /api/v1/carts/{}/coupon", id);
        CartResponse response = cartService.removeCoupon(id);
        return ResponseEntity.ok(response);
    }
}

File: services/nodejs-microservices/cart-service/src/infrastructure/web/CartController.ts
import { Router, Request, Response } from 'express';
import { CartServiceImpl } from '../../application/CartServiceImpl';
import { toCartSummaryResponse } from './dto';

export class CartController {
  public readonly router: Router;

  constructor(private readonly cartService: CartServiceImpl) {
    this.router = Router();
    this.setupRoutes();
  }

  private setupRoutes(): void {
    this.router.post('/carts', this.createCart.bind(this));
    this.router.get('/carts/:id', this.getCart.bind(this));
    this.router.post('/carts/:id/items', this.addItem.bind(this));
    this.router.put('/carts/:cartId/items/:itemId', this.updateItemQuantity.bind(this));
    this.router.delete('/carts/:cartId/items/:itemId', this.removeItem.bind(this));
    this.router.post('/carts/:id/coupon', this.applyCoupon.bind(this));
    this.router.delete('/carts/:id', this.clearCart.bind(this));
  }

  private async createCart(req: Request, res: Response): Promise<void> {
    try {
      if (!req.body.userId) { res.status(400).json({ error: 'userId is required' }); return; }
      const cart = await this.cartService.createCart(req.body);
      res.status(201).json(cart);
    } catch (error: any) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async getCart(req: Request, res: Response): Promise<void> {
    try {
      const summary = await this.cartService.getCartById(req.params.id);
      if (!summary) { res.status(404).json({ error: 'Cart not found' }); return; }
      res.json(toCartSummaryResponse(summary));
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async addItem(req: Request, res: Response): Promise<void> {
    try {
      if (!req.body.productId || !req.body.quantity) {
        res.status(400).json({ error: 'productId and quantity are required' }); return;
      }
      const item = await this.cartService.addItem(req.params.id, req.body);
      res.status(201).json(item);
    } catch (error: any) {
      if (error.message.includes('not found') || error.message.includes('not active')) {
        res.status(400).json({ error: error.message });
      } else {
        res.status(500).json({ error: 'Internal server error' });
      }
    }
  }

  private async updateItemQuantity(req: Request, res: Response): Promise<void> {
    try {
      if (!req.body.quantity) { res.status(400).json({ error: 'quantity is required' }); return; }
      const item = await this.cartService.updateItemQuantity(req.params.cartId, req.params.itemId, req.body.quantity);
      if (!item) { res.status(404).json({ error: 'Item not found' }); return; }
      res.json(item);
    } catch (error: any) {
      res.status(400).json({ error: error.message });
    }
  }

  private async removeItem(req: Request, res: Response): Promise<void> {
    try {
      const deleted = await this.cartService.removeItem(req.params.cartId, req.params.itemId);
      if (!deleted) { res.status(404).json({ error: 'Item not found' }); return; }
      res.status(204).send();
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async applyCoupon(req: Request, res: Response): Promise<void> {
    try {
      if (!req.body.couponCode) { res.status(400).json({ error: 'couponCode is required' }); return; }
      const cart = await this.cartService.applyCoupon(req.params.id, req.body);
      res.json(cart);
    } catch (error: any) {
      if (error.message.includes('Invalid') || error.message.includes('not found')) {
        res.status(400).json({ error: error.message });
      } else {
        res.status(500).json({ error: 'Internal server error' });
      }
    }
  }

  private async clearCart(req: Request, res: Response): Promise<void> {
    try {
      await this.cartService.clearCart(req.params.id);
      res.status(204).send();
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }
}
